cmd/k8s-operator/egress-services.go

```diff
@@ -0,0 +1,714 @@
+		tsoperator.SetServiceCondition(svc, tsapi.EgressSvcValid, metav1.ConditionUnknown, reasonProxyGroupNotReady, err.Error(), esr.clock, l)
+		return false, err
+	}
+	if !tsoperator.ProxyGroupIsReady(pg) {
+		l.Infof("ProxyGroup %s is not ready, waiting...", proxyGroupName)
+		tsoperator.SetServiceCondition(svc, tsapi.EgressSvcValid, metav1.ConditionUnknown, reasonProxyGroupNotReady, reasonProxyGroupNotReady, esr.clock, l)
+		return false, nil
+	}
+
+	if violations := validateEgressService(svc, pg); len(violations) > 0 {
+		msg := fmt.Sprintf("invalid egress Service: %s", strings.Join(violations, ", "))
+		esr.recorder.Event(svc, corev1.EventTypeWarning, "INVALIDSERVICE", msg)
+		l.Info(msg)
+		tsoperator.SetServiceCondition(svc, tsapi.EgressSvcValid, metav1.ConditionFalse, reasonEgressSvcInvalid, msg, esr.clock, l)
+		return false, nil
+	}
+	l.Debugf("egress service is valid")
+	tsoperator.SetServiceCondition(svc, tsapi.EgressSvcValid, metav1.ConditionTrue, reasonEgressSvcValid, reasonEgressSvcValid, esr.clock, l)
+	l.Debugf("egress service conditions %#+v", svc.Status)
+	return true, nil
+}
+
+func validateEgressService(svc *corev1.Service, pg *tsapi.ProxyGroup) []string {
+	violations := validateService(svc)
+	if len(svc.Spec.Ports) == 0 {
+		violations = append(violations, "egress Service for ProxyGroup must have at least one target Port specified")
+	}
+	if svc.Spec.Type != corev1.ServiceTypeExternalName {
+		violations = append(violations, fmt.Sprintf("unexpected egress Service type %s. The only supported type is ExternalName.", svc.Spec.Type))
+	}
+	if pg.Spec.Type != tsapi.ProxyGroupTypeEgress {
+		violations = append(violations, fmt.Sprintf("egress Service references ProxyGroup of type %s, must be type %s", pg.Spec.Type, tsapi.ProxyGroupTypeEgress))
+	}
+	return violations
+}
+
+// egressSvcNameBase returns a name base that can be passed to
+// ObjectMeta.GenerateName to generate a name for the ClusterIP Service.
+// The generated name needs to be short enough so that it can later be used to
+// generate a valid Kubernetes resource name for the EndpointSlice in form
+// 'ipv4-|ipv6-<ClusterIP Service name>.
+// A valid Kubernetes resource name must not be longer than 253 chars.
+func svcNameBase(s string) string {
+	// -ipv4 - ipv6
+	const maxClusterIPSvcNameLength = 253 - 5
+	base := fmt.Sprintf("ts-%s-", s)
+	generator := names.SimpleNameGenerator
+	for {
+		generatedName := generator.GenerateName(base)
+		excess := len(generatedName) - maxClusterIPSvcNameLength
+		if excess <= 0 {
+			return base
+		}
+		base = base[:len(base)-1-excess] // cut off the excess chars
+		base = base + "-"                // re-instate the dash
+	}
+}
+
+// unusedPort returns a port in range [3000 - 4000). The caller must ensure that
+// usedPorts does not contain all ports in range [3000 - 4000).
+func unusedPort(usedPorts sets.Set[int32]) int32 {
+	foundFreePort := false
+	var suggestPort int32
+	for !foundFreePort {
```

Comment:
I'm sure it's probably fine, but this algorithm's runtime will get a little bit choppy when there aren't many ports left, e.g. consider how lucky we need to get when there's only 1 remaining port. The expected number of iterations is 1000, but it's technically unbounded. Any reason not to just loop through all valid ports until we find one not in `usedPorts`? I don't see any downside as long as we can guarantee only one control loop is running at any one time (which we can for a specific reconcile target right?).

If we cared to do better than that, we could track the port number we got up to and start there next time. That would mean 1 iteration the first 1000 times, and then always starting with the port we allocated longest ago from then on. However, I don't think it's worth adding that complexity for the number of ports right now.

Reply:
> we could track the port number we got up to and start there next time.

Yes, I was thinking that when/if we will anyway need to store used ports on ProxyGroup status, we could also store the last used port there

> I'm sure it's probably fine, but this algorithm's runtime will get a little bit choppy when there aren't many ports left, e.g. consider how lucky we need to get when there's only 1 remaining port

Yeah, really we should figure out what the limits are (number of iptables rules, configmap size etc).
I generally prefer to restrict everything on start and loosen later if needed, so that we get more freedom to change things if needed without breaking anyone. For example, what if for some reasons users want to introduce the ability to define port ranges for network policy config reasons or if we need to set aside a port range for something else - it'd be nice to have a port range free for that
I can bump the 1000 up to something like 10000 though, that should probably make the possibility of running out of ports very unlikely + we can improve the algorithm once we set used ports on ProxyGroup status